### PR TITLE
fix stale hyperkit.pid making minikube start hang

### LINKED ISSUE #2994: hyperkit: start hangs after shutdown: SSH getsockopt: operation timed out (stale pid)
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

Please provide the following details:

**Environment**:

**Minikube version** (use `minikube version`): v0.28.1
- **OS** (e.g. from /etc/os-release): OSX High Sierra
- **VM Driver** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep DriverName`): hyperkit
- **ISO version** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep -i ISO` or `minikube ssh cat /etc/VERSION`): minikube-v0.28.0.iso
- **Install tools**: 
- **Others**:
The above can be generated in one go with the following commands (can be copied and pasted directly into your terminal):
```
minikube version
echo "";
echo "OS:";
cat /etc/os-release
echo "";
echo "VM driver": 
grep DriverName ~/.minikube/machines/minikube/config.json
echo "";
echo "ISO version";
grep -i ISO ~/.minikube/machines/minikube/config.json
```

**What happened**:

minikube start stuck at ssh after stop command or reboot my Mac.

**What you expected to happen**:

It'll be able to ssh and start the system.

**How to reproduce it** (as minimally and precisely as possible):

$ minikube start --vm-driver=hyperkit --memory 4096 --v 9 --alsologtostderr
wait until it running. Reboot Mac or $ minikube stop.
Run again
$ minikube start --vm-driver=hyperkit --memory 4096 --v 9 --alsologtostderr
Stuck
```
Waiting for SSH to be available...
Getting to WaitForSSH function...
(minikube) Calling .GetSSHHostname
(minikube) Calling .GetSSHPort
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHKeyPath
(minikube) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x14354b0] 0x1435460  [] 0s} 192.168.64.6 22 <nil> <nil>}
About to run SSH command:
exit 0


Error dialing TCP: dial tcp 192.168.64.6:22: getsockopt: operation timed out
```

**Output of `minikube logs` (if applicable)**:
hanging too.

**Anything else do we need to know**:
screen tty 
gives no output, so linuxkit is not running. 

Please let me know how to debug this.

### PR DESCRIPTION
**Pull request by balopat**
Fixes #2994.

### CODE REVIEW

**Comment by tstromberg** [comment]
```
--- FAIL: TestDocker (190.68s)
	util.go:246: got state: "(minikube) Found hyperkit pid: 12365\nRunning", expected "Running"
```

File: pkg/drivers/hyperkit/driver.go [hunk 1]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+
+	if _, err := os.Stat(pidFile); err != nil {
+		log.Infof("clean start, hyperkit pid file doesn't exist: %s", pidFile)
+		return nil
+	}
+
+	log.Warnf("hyperkit pid file exists: %s", pidFile)
+	// we have a pid file, hyperkit might be running!
+	// Make sure the pid written by hyperkit is the same as in the json
+	content, err := ioutil.ReadFile(pidFile)
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: Alternatively, since you are just reading one line and casting it to a string, you could also do:

r := bufio.NewReader(file)
line, err := r.ReadString('\n')

File: pkg/drivers/hyperkit/driver.go [hunk 2]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+	if err != nil {
+		return errors.Wrapf(err, "reading pidfile %s", pidFile)
+	}
+	pid, err := strconv.Atoi(string(content[:]))
+	if err != nil {
+		return errors.Wrapf(err, "parsing pidfile %s", pidFile)
+	}
+
+	p, _ := os.FindProcess(pid) //noop on linux
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: noop comment isn't useful here, since hyperkit is macOS only.

File: pkg/drivers/hyperkit/driver.go [hunk 3]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+	// Sending a signal of 0 can be used to check the existence of a process.
+	if err := p.Signal(syscall.Signal(0)); err == nil {
+		return fmt.Errorf("something is not right...please stop all minikube instances, seemingly a hyperkit server is already running with pid %d", pid)
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: Does it matter at all if this pid is hyperkit, and not a /bin/ls process that has reused the pid number?

If not, you might need to use a library like https://github.com/mitchellh/go-ps check that the process name contains hyperkit:

```
p := ps.FindProcess(pid)
if !strings.Contains(p.Executable(), "hyperkit") {
  .. stale pid file ...
}
```

File: pkg/drivers/hyperkit/driver.go [hunk 4]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
 	return nil
 }
 
+func (d *Driver) checkForPidFile() error {
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: checkForPidFile sounds like a function that simply asserts its existence. Since this does a bit more, please consider an alternative name and/or add a comment describing the goal of this function.

File: pkg/drivers/hyperkit/driver.go [hunk 5]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+	pidFile := filepath.Join(stateDir, pidFileName)
+
+	if _, err := os.Stat(pidFile); err != nil {
+		log.Infof("clean start, hyperkit pid file doesn't exist: %s", pidFile)
+		return nil
+	}
+
+	log.Warnf("hyperkit pid file exists: %s", pidFile)
+	// we have a pid file, hyperkit might be running!
+	// Make sure the pid written by hyperkit is the same as in the json
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: This comment would be better placed lower down in the function, where you read d.Pid.

File: pkg/drivers/hyperkit/driver.go [hunk 6]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+	p, _ := os.FindProcess(pid) //noop on linux
+	// Sending a signal of 0 can be used to check the existence of a process.
+	if err := p.Signal(syscall.Signal(0)); err == nil {
+		return fmt.Errorf("something is not right...please stop all minikube instances, seemingly a hyperkit server is already running with pid %d", pid)
+	}
+
+	machinePID := d.getPid()
+
+	if machinePID != pid {
+		return fmt.Errorf("something is not right...the PID reported in %s (%d) is different from machine PID (%d from %s)", pidFile, pid, machinePID, machineFileName)
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: Would it be possible to signal how users should recover from this state mismatch?

File: pkg/drivers/hyperkit/driver.go [hunk 7]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+	}
+
+	log.Warnf("No running process found with PID %d, removing %s...", machinePID, pidFile)
+	if err := os.Remove(pidFile); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("removing pidFile %s", pidFile))
+	}
+	log.Infof("Removed PID file.")
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: Remove superfluous comment.

**Comment by tstromberg** [comment]
FYI, this change is wonderful and I can't wait to see it merged. Thank you!

File: pkg/drivers/hyperkit/driver.go [hunk 8]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
...
+func (d *Driver) checkForPidFile() error {
+	stateDir := filepath.Join(d.StorePath, "machines", d.MachineName)
+	pidFile := filepath.Join(stateDir, pidFileName)
+
+	if _, err := os.Stat(pidFile); err != nil {
+		log.Infof("clean start, hyperkit pid file doesn't exist: %s", pidFile)
+		return nil
+	}
+
+	log.Warnf("hyperkit pid file exists: %s", pidFile)
```
[the remark this answers is not in the document]
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: I do want to warn the user that there was an unclean shutdown - even though we are betting based on experience that "just removing the pid file" is a solution in most cases, but unclean shutdown/restart of a VM still might result in unexpected situations. Maybe rewording would help?

File: pkg/drivers/hyperkit/driver.go [hunk 9]
```diff
@@ -240,6 +245,48 @@ func (d *Driver) Start() error {
 	return nil
 }
 
+func (d *Driver) checkForPidFile() error {
+	stateDir := filepath.Join(d.StorePath, "machines", d.MachineName)
+	pidFile := filepath.Join(stateDir, pidFileName)
+
+	if _, err := os.Stat(pidFile); err != nil {
```
[the remark this answers is not in the document]
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: ack

File: pkg/drivers/hyperkit/driver.go (hunk 2 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: ack - copypaste error

File: pkg/drivers/hyperkit/driver.go (hunk 3 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: It does matter - as if there is a hyperkit instance with this same exact PID - as there is a non-zero chance that we are looking at some kind of race condition with another starting minikube instance with the same name and then we should abort. There is also a chance it's a different minikube instance with different name - but then again - I can't really differentiate between the two so I'd rather fail here. 

If the PID is reused by another non-hyperkit instance, then I don't care, I can actually remove the pid file and have hyperkit restart with another PID.

File: pkg/drivers/hyperkit/driver.go (hunk 6 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: I think I'll remove this consistency check - not sure what scenarios it helps with: maybe some cases of manual tempering with the pid file? Even then it's not a great check as someone can just simply modify the json too. I think this is unnecessary.

File: pkg/drivers/hyperkit/driver.go (hunk 7 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: ack

File: pkg/drivers/hyperkit/driver.go (hunk 2 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: oh, actually it should be noop on UNIX :)

File: pkg/drivers/hyperkit/driver.go (hunk 1 above)
Comment (balopat) [reply pkg/drivers/hyperkit/driver.go]: Why is this better? It doesn't win in number of lines as then I'll need to create & close the file myself, ioutil.ReadFile manages for me :)

**Comment by balopat** [comment]
> Is some additional work required here?
> 
> ```
> --- FAIL: TestDocker (190.68s)
> 	util.go:246: got state: "(minikube) Found hyperkit pid: 12365\nRunning", expected "Running"
> ```

That was a bug that an additional debugging line introduced! I removed the offending log.

File: pkg/drivers/hyperkit/driver.go [hunk 10]
```diff
@@ -183,13 +189,20 @@ func (d *Driver) Start() error {
...
 		h.VSockPorts = vsockPorts
 	}
+	h.Disks = []hyperkit.DiskConfig{
+		{
+			Path:   pkgdrivers.GetDiskPath(d.BaseDriver),
+			Size:   d.DiskSize,
+			Driver: "virtio-blk",
+		},
+	}
+	//}
```
Comment (tstromberg) [inline pkg/drivers/hyperkit/driver.go]: Remove comment

**Comment by tstromberg** [comment]
Looks great. Please merge once test failures are addressed:

```
 15:47:58 | > (minikube) clean start, hyperkit pid file doesn't exist: /Users/jenkins/minikube-integration/darwin-amd64-hyperkit-3593-77386-5af2f1eafc7a5d00238d05637682bf4264343497/.minikube/machines/minikube/hyperkit.pid
 15:47:58 | > (minikube) Using UUID 23979220-2357-11e9-ac78-38c98626f627
 15:48:00 | ! E0128 15:48:00.546915   77513 start.go:205] Error starting host: Error creating host: Error creating machine: Error in driver during machine creation: getting MAC address from UUID: vmnet: error from vmnet.framework: -1
 15:48:00 | ! .
 15:48:00 | !  Retrying.
 15:48:00 | ! I0128 15:48:00.546971   77513 utils.go:114] non-retriable error: Error creating host: Error creating machine: Error in driver during machine creation: getting MAC address from UUID: vmnet: error from vmnet.framework: -1
 15:48:00 | > minikube failed :( exiting with error code 1
 15:48:00 | ! E0128 15:48:00.546993   77513 start.go:211] Error starting host:  Error creating host: Error creating machine: Error in driver during machine creation: getting MAC address from UUID: vmnet: error from vmnet.framework: -1
```

**Comment by balopat** [comment]
in fact there is a running hyperkit instance: 

```
55199   ??  S     18:06.13 /usr/local/bin/hyperkit -A -u -F /Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/hyperkit.pid -c 2 -m 2048M -s 0:0,hostbridge -s 31,lpc -s 1:0,virtio-net -U 304f5632-22ae-11e9-846f-38c98626f627 -s 2:0,virtio-blk,/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/minikube.rawdisk -s 3,ahci-cd,/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/boot2docker.iso -s 4,virtio-rnd -l com1,autopty=/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/tty,log=/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/console-ring -f kexec,/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/bzimage,/Users/jenkins/minikube-integration/darwin-amd64-hyperkit-master-54872-00be38dadc7c4301471c9d6657853252bc729e4a/.minikube/machines/minikube/initrd,earlyprintk=serial loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10 systemd.legacy_systemd_cgroup_controller=yes base host=minikube
```
I'm going to kill it and retry the job.

**Comment by balopat** [comment]
I wonder where the collision was - one candidate is `host=minikube` - and would make sense that virtio-net wouldn't allow two hosts with the same name...

**Comment by tstromberg** [comment]
@minikube-bot OK to test
